File: python/day20.py
from collections import defaultdict
import re


with open('./inputs/20.txt') as f:
    instructions = f.readlines()

def magnitude(particle_accelerations):
    return sum(r**2 for r in particle_accelerations)

pattern = re.compile(r'p=<(-*\d+,-*\d+,-*\d+)>, v=<(-*\d+,-*\d+,-*\d+)>, a=<(-*\d+,-*\d+,-*\d+)>')
particles = []
smallest_acceleration = 98765
closest_particle = None

for i, line in enumerate(instructions):
    particle = tuple(tuple(map(int, g.split(','))) for g in re.match(pattern, line).groups())
    accel = magnitude(particle[-1])
    if accel < smallest_acceleration:
        smallest_acceleration = accel
        closest_particle = i
    particles.append(particle)

print(closest_particle)




def get_differences(x, y):
    diff = lambda r_x, r_y: (r_x[0] - r_y[0], r_x[1] - r_y[1], r_x[2] - r_y[2])
    p_x, v_x, a_x = x
    p_y, v_y, a_y = y

    p_diffs = diff(p_x, p_y)
    v_diffs = diff(v_x, v_y)
    a_diffs = diff(a_x, a_y)
    return p_diffs, v_diffs, a_diffs


def check_collision(ppp, vvv, aaa, t):
    return all(0.5*a*t*t + (v+0.5*a)*t + p == 0 for p, v, a in zip(ppp, vvv, aaa))


def find_collisions(x, y):
    ppp, vvv, aaa = get_differences(x, y)
    p, v, a = ppp[0], vvv[0], aaa[0]
    b = -v - 0.5 * a
    D = b*b - 2 * a * p
    if a == 0:
        if v != 0:
            t = -p/v
            if check_collision(ppp, vvv, aaa, t):
                return t
    elif D == 0:
        t = b/a
        if check_collision(ppp, vvv, aaa, t):
            return t
    else:
        for t in ((b - D**0.5)/a, (b + D**0.5)/a):
            if check_collision(ppp, vvv, aaa, t):
                return t
    return 0


collisions = defaultdict(set)
for i, p1 in enumerate(particles):
    for j, p2 in enumerate(particles[i+1:], i+1):
        time = find_collisions(p1, p2)
        if time > 0:
            collisions[time].add((i, j))

alive = set(range(len(particles)))
for time in sorted(collisions.keys()):
    now_colliding = set()
    for p1, p2 in collisions[time]:
        if p1 in alive and p2 in alive:
            now_colliding |= {p1, p2}
    alive -= now_colliding

print(len(alive))

File: nim/day20.nim
import strutils, sequtils, tables, math


const instructions = readFile("./inputs/20.txt").splitLines
type
  Coord = tuple[x, y, z: float]
  Particle = tuple[p, v, a: Coord]

func extractNumbers(s: string): Coord =
  let xyz = s[3 .. ^2].split(',').map(parseFloat)
  (xyz[0], xyz[1], xyz[2])

func parseLine(line: string): Particle =
  let pva = line.split(", ").map(extractNumbers)
  (pva[0], pva[1], pva[2])

func manhattan(accel: Coord): float =
  abs(accel.x) + abs(accel.y) + abs(accel.z)

var
  particles: array[instructions.len, Particle]
  smallestAcceleration = float.high
  closestParticle: int

for i, line in instructions:
  let
    particle = line.parseLine()
    ma = manhattan(particle.a)
  if ma < smallestAcceleration:
    smallestAcceleration = ma
    closestParticle = i
  particles[i] = particle

echo closestParticle




func `-`(c1, c2: Coord): Coord = (c1.x - c2.x, c1.y - c2.y, c1.z - c2.z)
func `-`(p1, p2: Particle): Particle = (p1.p - p2.p, p1.v - p2.v, p1.a - p2.a)

func collision(pp, vv, aa, time: float): bool =
    (2 * vv + aa * (1+time)) * time + 2 * pp == 0

func checkCollision(p, v, a: Coord, time: float): bool =
  collision(p.y, v.y, a.y, time) and collision(p.z, v.z, a.z, time)

func findCollisions(p1, p2: Particle): float =
  let
    (p, v, a) = p1 - p2
    b = - v.x - 0.5 * a.x
    D = b*b - 2 * a.x * p.x
  if a.x == 0:
    if v.x != 0:
      let time = -p.x / v.x
      if checkCollision(p, v, a, time):
        return time
  elif D == 0:
    let time = b / a.x
    if checkCollision(p, v, a, time):
      return time
  else:
    for time in [(b - D.sqrt) / a.x, (b + D.sqrt) / a.x]:
      if checkCollision(p, v, a, time):
        return time


var collisions = initTable[int, seq[tuple[i, j: int16]]]()

for i, p1 in particles:
  for j, p2 in particles[i+1 .. particles.high]:
    let time = findCollisions(p1, p2).int
    if time > 0:
      if not collisions.hasKey(time):
        collisions[time] = @[]
      collisions[time].add((i.int16, (i+j+1).int16))

var dead: set[int16]
for time in collisions.keys:
  var colliding: set[int16]
  for collision in collisions[time]:
    if collision.i notin dead and collision.j notin dead:
      colliding.incl(collision.i)
      colliding.incl(collision.j)
  dead = dead + colliding

echo instructions.len - dead.card
